src/ProducerEventProcessor/UploaderCaptureEventCollectorTest.cpp

```diff
@@ -0,0 +1,152 @@
+  }
+
+ protected:
+  void TearDown() override {
+    absl::MutexLock lock{&mutex_};
+    CHECK(stop_called_);
+
+    if (data_produce_thread_.joinable()) data_produce_thread_.join();
+    if (data_upload_thread_.joinable()) data_upload_thread_.join();
+  }
+
+  UploaderCaptureEventCollector collector_;
+  size_t total_uploaded_data_bytes_ = 0;
+  uint64_t wait_all_data_uploaded_duration_ms_;
+
+ private:
+  // Keep producing capture event until stop is called or `event_count_` events are produced with a
+  // specified produce duration. The last produced event will always be the capture finish event.
+  void ProduceCaptureData() {
+    size_t event_count = event_count_;
+    bool stopped = false;
+    while (event_count-- > 1 && !stopped) {
+      mutex_.LockWhenWithTimeout(absl::Condition(&stop_called_),
+                                 absl::Milliseconds(produce_duration_ms_));
+      collector_.AddEvent(CreateInternedStringCaptureEvent(kAnswerKey, kAnswerString));
+      stopped = stop_called_;
+      mutex_.Unlock();
+    }
+
+    collector_.AddEvent(CreateCaptureFinishedEvent());
+
+    Stop();
+  }
+
+  void UploadCaptureData() {
+    const std::vector<unsigned char>& upload_data_buffer_ = collector_.GetUploadDataBuffer();
+
+    while (collector_.GetDataReadiness() != orbit_capture_uploader::DataReadiness::kEndOfData) {
+      std::this_thread::sleep_for(std::chrono::milliseconds(upload_duration_ms_));
+      total_uploaded_data_bytes_ += upload_data_buffer_.size();
+      collector_.RefreshUploadDataBuffer();
+    }
+  }
+
+  absl::Mutex mutex_;
+  bool stop_called_ ABSL_GUARDED_BY(mutex_) = false;
+  size_t event_count_ = 0;
+  uint64_t produce_duration_ms_;
+  uint64_t upload_duration_ms_;
+  std::thread data_upload_thread_;
+  std::thread data_produce_thread_;
+};
+
+}  // namespace
+
+TEST_F(UploaderCaptureEventCollectorTest, EarlyStop) {
+  constexpr uint64_t kEventCount = 10;
+  constexpr uint64_t kProduceDurationMs = 50;
+  constexpr uint64_t kUploadDurationMs = 5;
+  constexpr uint64_t kEarlyStopMs = 120;
+  constexpr uint64_t kActualProducedEventCount = 4;
+  Start(kEventCount, kProduceDurationMs, kUploadDurationMs);
+
+  // Stop producing data early and wait until all data is uploaded
+  std::this_thread::sleep_for(std::chrono::milliseconds(kEarlyStopMs));
+  Stop();
+  std::this_thread::sleep_for(
+      std::chrono::milliseconds(wait_all_data_uploaded_duration_ms_ - kEarlyStopMs));
+  EXPECT_EQ(collector_.GetTotalUploadedEventCount(), kActualProducedEventCount);
+  EXPECT_EQ(collector_.GetTotalUploadedDataBytes(), total_uploaded_data_bytes_);
+}
+
+TEST_F(UploaderCaptureEventCollectorTest, SlowProduceAndFastUpload) {
+  constexpr uint64_t kEventCount = 10;
+  constexpr uint64_t kSlowProduceDurationMs = 50;
+  constexpr uint64_t kFastUploadDurationMs = 5;
+  Start(kEventCount, kSlowProduceDurationMs, kFastUploadDurationMs);
+
+  std::this_thread::sleep_for(std::chrono::milliseconds(wait_all_data_uploaded_duration_ms_));
```

Comment:
Same as before. It would be better to wait for a mutex instead of sleeping.

Reply:
Done! Thanks!